Replace Bluebird with Promise - next stage

liranmauda opened the pull request:
### Explain the changes
- Using delay in a native way (new delay function at promise_utils)
- Removed all `.tap()`, `.spread()` and `.return()`
- Removed most of `P.join`
- Wrapped with bluebird some of the timeout and catch

### Out of scope
- Did not replaced `.timeout()`
- Did not replaced `.cancel()`
- Did not replaced `.isPending()`
- Did not replaced `.isRejected()`
- Did not replaced `.props()`
- Did not replaced 2 `.delay()`

Step 1 for #6070 

Signed-off-by: liranmauda <liran.mauda@gmail.com>

guymguym reviewed (changes requested):
@liranmauda That's a good progress indeed! 

But since you are replacing a bunch of functions that used to return Bluebird's to return native Promise instead, we risk that callers that depend on Bluebird instance methods will fail because a native Promise is returned somewhere in the middle of the promise chain.

So I think with this PR we also need to go over \*\*all\*\* the calls to Bluebird methods and either eliminate them or make sure they are scoped to a locally constructed Bluebird promise. Notice that we don't need to go over static functions like P.fromCallback, only methods that exist on a Bluebird instance) - 

```javascript
> native_methods = Object.getOwnPropertyNames(Promise.prototype)
[ 'constructor', 'then', 'catch', 'finally' ]

> Object.getOwnPropertyNames(require('bluebird').prototype).filter(n => !n.startsWith('_') && !native_methods.includes(n))
[
  'suppressUnhandledRejections',
  'lastly',
  'tap',
  'tapCatch',
  'toString',
  'caught',
  'reflect',
  'done',
  'spread',
  'toJSON',
  'all',
  'error',
  'bind',
  'cancel',
  'break',
  'isCancellable',
  'thenReturn',
  'return',
  'thenThrow',
  'throw',
  'catchThrow',
  'catchReturn',
  'isCancelled',
  'isPending',
  'isRejected',
  'isFulfilled',
  'isResolved',
  'value',
  'reason',
  'call',
  'get',
  'map',
  'nodeify',
  'asCallback',
  'props',
  'race',
  'reduce',
  'settle',
  'some',
  'delay',
  'timeout',
  'disposer',
  'any',
  'each',
  'mapSeries',
  'filter'
]
```

guymguym commented:
This gets the list 
```bash
$ RE=$(node -p '
  Object.keys(require("bluebird").prototype)
  .filter(n => !n.startsWith("_") && ![
    "then", "catch", "finally",                       // Promise
    "filter", "map", "reduce", "some", "any", "all",  // Array
    "each", "get",                                    // lodash
    "error",                                          // console/dbg
    "call", "bind",                                   // Function
    "toString", "toJSON",                             // Object
  ].includes(n))
  .map(n=>`\\.${n}\\(`)
  .join("|")
')
$ rg \'$RE\' src | egrep -v '^src/test/' | wc -l
     161
$ rg --only-matching \'$RE\' src | egrep -v '^src/test/' | cut -d':' -f2- | sort | uniq -c | sort -n
   1 .cancel(
   1 .isPending(
   1 .isRejected(
   2 .props(
  10 .tap(
  12 .spread(
  37 .timeout(
  47 .return(
  50 .delay(
```
So most of the bluebird methods we use are `delay, return, timeout, spread, tap...` etc.
If you eliminate those 160 callers it will become much easier to replace bluebird promises with native promises because we won't worry that callers may fail.

BTW 1 - notice that I filtered out `src/tests/` assuming that we should detect test errors by failures.
BTW 2 - I filtered quite a lot of known function names that collide with other objects such as arrays, so it might still be that we used one of those bluebird functions somewhere, but that's not very likely.

liranmauda commented:
> If you eliminate those 160 callers it will become much easier to replace bluebird promises with native promises because we won't worry that callers may fail.

by eliminate you mean  do not replace to native where we call those functions (or functions that return bluebird that are chained)?

liranmauda commented:
.cancel() - chained to ClientRequest which we didn’t touced. 
.isPending() - chained to RpcRequest which we didn’t touced. 
.isRejected() - doesn't look like we touched.
.props() - all is chained to P. 
.tap():

Being called here: 
https://github.com/noobaa/noobaa-core/blob/master/src/endpoint/ftp/ftp_filesystem.js#L93
And here:
https://github.com/noobaa/noobaa-core/blob/master/src/endpoint/ftp/ftp_filesystem.js#L126
Which are not bluebird (on master) 

.spread() - is called mostly from P.join and once from P.fromCallback
all the other was looked at the first commit :) 
Hope we didn't missed anything.

guymguym commented:
> by eliminate you mean do not replace to native where we call those functions (or functions that return bluebird that are chained)?

@liranmauda No I meant that we need to remove those references to Bluebird methods so that we are more certain that we can't crash on that.

I think we just need to replace those with some other simple one liners or utility, for example:
```
Promise.resolve()
    .then(() => 42)
    .delay(1000)
    .then(console.log)
```
can be replaced to 
```
Promise.resolve()
    .then(() => 42)
    .then(promise_utils.delay(1000))
    .then(console.log)
```

and get rid also from: .spread() .timeout() .return().

liranmauda commented:
- Removed all .tap() and .return() 
- Removed most of P.join 
- Removed most of .spread()

liranmauda committed:
Replace Bluebird with Promise - next stage

- Replace P.join, which is deprecated, with Promise.all([])
- Replace new P with new Promise
- Add delay function into promise_ustils.js
- Remove tap
- Replaced return with value and empty return
- Replaced .spread()
- Wrapping with Bluebird special cases

Signed-off-by: liranmauda <liran.mauda@gmail.com>

guymguym commented:
You should probably remove all of .spread calls because it will fail on native promises and is risky to leave it, and we can easily replace it with array param destructoring..

guymguym commented:
What about .timeout?

liranmauda commented:
> You should probably remove all of .spread calls because it will fail on native promises and is risky to leave it, and we can easily replace it with array param destructoring..

Yes, I removed most in some of the commits but I fail on unit tests (only on travis) so I am removing and adding commits.
In later commit i left only 4 .spread 

> What about .timeout? we can replace it with a promise_utils and get rid of those bluebird calls

Still didn't figure out how to replace .timeout

liranmauda commented:
removed the delay chang, Will add it once I figure out why tarvis fails on it :)

guymguym commented:
BTW this also means to change those catch timeout such as:
https://github.com/noobaa/noobaa-core/blob/9b6ec6bf3263d2cb531ec2237f73a0128c9bf184/src/rpc/rpc_base_conn.js#L128-L135

because this is again a non-standard catch variant of bluebird.

so with promise_utils.timeout we can write:

```javascript
try {
  const encoded_message = this._encode_message(msg);
  const send_promise = this._send(encoded_message, msg.body && msg.body.op, req);
  const res = await promise_utils.timeout(config.RPC_SEND_TIMEOUT, send_promise);
  return res;
} catch(err) {
  if (err instanceof promise_utils.TimeoutError) {
    this.emit_error(new RpcError('RPC_SEND_TIMEOUT', 'RPC SEND TIMEOUT'));
  } else {
    this.emit_error(err);
  }
}
```

liranmauda commented on src/agent/agent_cli.js:
```diff
@@ -329,7 +329,7 @@ class AgentCLI {
             return child_process.execAsync('attrib +H ' + current_path)
-                .then(() => P.join(os_utils.is_folder_permissions_set(current_path), fs.readdirAsync(current_path)))
+                .then(() => Promise.all([os_utils.is_folder_permissions_set(current_path), fs.readdirAsync(current_path)]))
                 .spread(function(permissions_set, noobaa_storage_initialization) {
```
Converted all the `.spreds()` into this: 
```
.then(function(res) {
                    let permissions_set = res[0];
                    let noobaa_storage_initialization = res[1];
...
```
should I refractor it to just `.then( ([ arg1, arg2 ]) => { ... })`?

guymguym replied:
yes - .then( ([ arg1, arg2 ]) => { ... })

guymguym commented on src/util/promise_utils.js:
```diff
@@ -368,7 +381,9 @@ function map_values(obj, func) {
             })
         ))
-        .return(new_obj);
+        .then(function() {
+            return new_obj;
+        });
```
```suggestion
        .then(() => new_obj);
```

liranmauda replied:
sure, I will do it for all those entries.

guymguym commented on src/agent/agent_diagnostics.js:
```diff
@@ -68,7 +68,9 @@ function collect_agent_diagnostics(storage_path) {
         .finally(() => diag_log.end())
-        .return()
+        .then(() => {
+            // do nothing. 
```
I think it's important to note that we "return undefined" and not that we do nothing.
It is even better be explicit like `.then(() => undefined)`

liranmauda replied:
sure, I will do it for all those entries.

guymguym commented on src/rpc/ice.js:
```diff
@@ -1496,11 +1496,13 @@ function listen_on_port_range(port_range) {
         // wait for listen even, while also watching for error/close.
         return promise_utils.wait_for_event(server, 'listening')
-            .return(server)
+            .then(function() {
+                return server;
+            })
```
you don't want to define functions in promise chains because any access to `this` will not work, so prefer arrow functions in chains:
```suggestion
            .then(() => server)
```

guymguym commented on src/endpoint/ftp/ftp_filesystem.js:
```diff
@@ -90,7 +90,10 @@ class FtpFileSystemNB {
                 limit: 100
             }))
-            .tap(console.log)
+            .then(res => {
+                console.log(res);
+                return res;
+            })
             .then(res => {
                 console.log(`in get - got res = `, res);
```
just delete that then because the next one prints anyhow

guymguym commented on src/agent/block_store_services/block_store_fs.js:
```diff
@@ -61,16 +61,18 @@ class BlockStoreFs extends BlockStoreBase {
-            )
-            .spread((usage, drive) => {
+            ])
+            .then(res => {
+                let usage = res[0];
+                let drive = res[1];
```
```suggestion
            .then(([usage, drive]) => {
```

guymguym commented on src/agent/block_store_services/block_store_fs.js:
```diff
@@ -218,7 +220,10 @@ class BlockStoreFs extends BlockStoreBase {
                 let usage_data = JSON.stringify(this._usage);
-                return fs.writeFileAsync(this.usage_path, usage_data).return(usage);
+                return fs.writeFileAsync(this.usage_path, usage_data)
+                    .then(function() {
+                        return usage;
+                    });
```
use arrow
```suggestion
                    .then(() => usage);
```

guymguym commented on src/rpc/rpc.js:
```diff
@@ -904,7 +904,10 @@ class RPC extends EventEmitter {
         tcp_server.on('connection', conn => this._accept_new_connection(conn));
-        return P.resolve(tcp_server.listen(port)).return(tcp_server);
+        return Promise.resolve(tcp_server.listen(port))
+            .then(function() {
+                return tcp_server;
+            });
```
prefer arrow
```suggestion
            .then(() => tcp_server);
```

guymguym commented on src/rpc/rpc.js:
```diff
@@ -917,7 +920,10 @@ class RPC extends EventEmitter {
         ntcp_server.on('connection', conn => this._accept_new_connection(conn));
-        return P.resolve(ntcp_server.listen(port)).return(ntcp_server);
+        return Promise.resolve(ntcp_server.listen(port))
+            .then(function() {
+                return ntcp_server;
+            });
```
```suggestion
            .then(() => ntcp_server);
```

guymguym commented on src/rpc/rpc_http_server.js:
```diff
@@ -53,7 +53,9 @@ class RpcHttpServer extends events.EventEmitter {
         this.install_on_server(server, options.default_handler);
         return P.fromCallback(callback => server.listen(port, callback))
-            .return(server);
+            .then(function() {
+                return server;
+            });
```
arrow...

guymguym commented on src/rpc/stun.js:
```diff
@@ -617,7 +618,7 @@ function test() {
                         P.ninvoke(socket, 'send', req, 0, req.length, stun_url.port, stun_url.hostname),
                         P.ninvoke(socket, 'send', ind, 0, ind.length, stun_url.port, stun_url.hostname))
-                    .delay(stun.INDICATION_INTERVAL * chance.floating(stun.INDICATION_JITTER))
+                    .then(promise_utils.delay(stun.INDICATION_INTERVAL * chance.floating(stun.INDICATION_JITTER)))
```
OK this is a pitfall.
You must call the delay only when you actually need to start waiting, which is inside the promise chain, and not from the outside function. So please make sure to always use like this `.then(() => promise_utils.delay(555))`

```suggestion
                    .then(() => promise_utils.delay(stun.INDICATION_INTERVAL * chance.floating(stun.INDICATION_JITTER)))
```

guymguym replied:
BTW notice that for all the cases you use it with await there is no need to use an arrow function because that's only needed for `.then(() => ...)`

guymguym commented on src/server/analytic_services/activity_log_store.js:
```diff
@@ -44,13 +44,15 @@ class ActivityLogStore {
             .then(() => this._activitylogs.col().insertOne(activity_log))
             .catch(err => mongo_utils.check_duplicate_key_conflict(err, 'audit_log'))
-            .return(activity_log);
+            .then(function() {
+                return activity_log;
+            });
```
arrow
```suggestion
            .then(() => activity_log);
```

guymguym commented on src/server/bg_services/agent_blocks_verifier.js:
```diff
@@ -108,7 +108,9 @@ class AgentBlocksVerifier {
             })
-            .return();
+            .then(() => {
+                // do nothing. 
```
```suggestion
                // return nothing. 
```

guymguym commented on src/server/bg_services/cluster_hb.js:
```diff
@@ -60,8 +60,12 @@ function do_heartbeat({ skip_server_monitor } = {}) {
-            .spread((mongo_status, drives, raw_storage) => {
+                os_utils.get_raw_storage()
+            ]))
+            .then(res => {
+                let mongo_status = res[0];
+                let drives = res[1];
+                let raw_storage = res[2];
```
```suggestion
            .then(([mongo_status, drives, raw_storage]) => {
```

guymguym commented on src/server/bg_services/cluster_hb.js:
```diff
@@ -119,11 +123,13 @@ function do_heartbeat({ skip_server_monitor } = {}) {
             })
-            .return();
+            .then(() => {
+                // do nothing. 
```
```suggestion
                // return nothing. 
```

guymguym commented on src/server/bg_services/md_aggregator.js:
```diff
@@ -254,21 +254,21 @@ function range_md_aggregator({
-            existing_blocks_aggregate,
-            deleted_blocks_aggregate) => {
+        ])
+        .then(res => {
```
.then(([
 ...
]) => {

liranmauda commented:
now we left with:
```
$ rg --only-matching \'$RE\' src | egrep -v '^src/test/' | cut -d':' -f2- | sort | uniq -c | sort -n
   1 .cancel(
   1 .isPending(
   1 .isRejected(
   1 .spread(
   2 .props(
  37 .timeout(
  46 .delay(
```
where delay number is way less (guess 2), as it counts it wrongly :)

guymguym commented on src/util/promise_utils.js:
```diff
@@ -62,7 +64,8 @@ function iterate(array, func) {
     }
 
-    return P.try(next).return(results);
+    return P.try(next)
+        .then(() => results);
```
```suggestion
    return Promise.resolve().then(next).then(() => results);
```

guymguym replied:
we don't want bluebird here..

liranmauda replied:
sure, in that function we still have 2 blubirds:
https://github.com/noobaa/noobaa-core/blob/9b6ec6bf3263d2cb531ec2237f73a0128c9bf184/src/util/promise_utils.js#L39
and 
https://github.com/noobaa/noobaa-core/blob/9b6ec6bf3263d2cb531ec2237f73a0128c9bf184/src/util/promise_utils.js#L62

should I leave them?

guymguym commented on src/server/system_services/cluster_server.js:
```diff
@@ -186,13 +186,16 @@ function add_member_to_cluster_invoke(req, my_address) {
+            ]);
         })
-        .spread((root_ca, server_cert, client_cert) => {
+        .then(res => {
+            let root_ca = res[0];
+            let server_cert = res[1];
+            let client_cert = res[2];
```
```suggestion
        .then(([root_ca, server_cert, client_cert])=> {
```

guymguym commented on src/server/system_services/stats_aggregator.js:
```diff
@@ -569,7 +569,9 @@ function get_nodes_stats(req) {
                 })
             }))
-        .spread((nodes_results, hosts_results) => {
+        .then(res => {
+            let nodes_results = res[0];
+            let hosts_results = res[1];
```
```suggestion
        .then(([nodes_results, hosts_results])=> {
```

liranmauda replied:
when doing that we get a lint error:
`Unexpected block statement surrounding arrow body; move the returned value immediately after the =>.eslint(arrow-body-style)`

guymguym commented on src/server/system_services/tier_server.js:
```diff
@@ -101,11 +100,13 @@ function read_tier(req) {
-        )
-        .spread(function(nodes_aggregate_pool, available_to_upload) {
+        ])
+        .then(function(res) {
+            let nodes_aggregate_pool = res[0];
+            let available_to_upload = res[1];
```
```suggestion
        .then(([nodes_aggregate_pool, available_to_upload]) {
```

liranmauda replied:
when doing that we get a lint error:
`Unexpected block statement surrounding arrow body; move the returned value immediately after the =>.eslint(arrow-body-style)`

guymguym replied:
I was just missing the arrow `=>` Fixed the suggestion

liranmauda replied:
@guymguym not it, I manually added the arrow and got this lint error
so the code is:
```
        .then(([nodes_aggregate_pool, available_to_upload]) => {
            return get_tier_info(tier, nodes_aggregate_pool, { mirror_storage: available_to_upload[String(tier._id)] });
        });
```
but lint doesn't like it

guymguym replied:
because the linter is asking you to get rid of the return like this:
```javascript
        .then(([nodes_aggregate_pool, available_to_upload]) =>
            get_tier_info(tier, nodes_aggregate_pool, { mirror_storage: available_to_upload[String(tier._id)] })
        );
```

guymguym commented on src/server/system_services/tier_server.js:
```diff
@@ -538,11 +539,13 @@ function read_policy(req) {
-        )
-        .spread(function(nodes_aggregate_pool, hosts_aggregate_pool) {
+        ])
+        .then(function(res) {
+            let nodes_aggregate_pool = res[0];
+            let hosts_aggregate_pool = res[1];
```
```suggestion
        .then(([nodes_aggregate_pool, hosts_aggregate_pool]) {
```

liranmauda replied:
when doing that we get a lint error:
`Unexpected block statement surrounding arrow body; move the returned value immediately after the =>.eslint(arrow-body-style)`

guymguym replied:
I was just missing the arrow `=>` Fixed the suggestion

nixpanic commented:
/retest ci/centos/unit

nimrod-becker commented:
/retest ci/centos/unit

liranmauda commented:
/retest ci/centos/unit

liranmauda commented:
/retest ci/centos/unit

guymguym commented on src/agent/agent.js:
```diff
@@ -700,11 +704,12 @@ class Agent {
-                    this._test_connection_timeout = null;
-                })
+            P.resolve(
+                    this.client.node.ping()
+                    .timeout(MASTER_RESPONSE_TIMEOUT)
```
```suggestion
                    Bluebird.resolve(this.client.node.ping())
                        .timeout(MASTER_RESPONSE_TIMEOUT)
```

guymguym commented:
@liranmauda Let me know when you rebase from master and push the updated code.

guymguym commented on src/util/promise_utils.js:
```diff
@@ -63,7 +65,7 @@ function iterate(array, func) {
         return P.try(() => func(array[i], i, array)).then(next);
```
What about this P.try?

guymguym commented on src/util/promise_utils.js:
```diff
@@ -5,11 +5,13 @@ const _ = require('lodash');
 
 require('setimmediate');
 
+const setTimeoutAsync = util.promisify(setTimeout);
```
so how about just defining
```
const delay = util.promisify(setTimeout);
```

guymguym commented on src/util/promise_utils.js:
```diff
@@ -169,18 +179,18 @@ function next_tick() {
 // so that it won't prevent the process from existing if it's the only timer left
 function run_background_worker(worker) {
-    var DEFUALT_DELAY = 10000;
+    var DEFAULT_DELAY = 10000;
```
const

guymguym commented on src/agent/block_store_services/block_store_fs.js:
```diff
@@ -92,7 +92,7 @@ class BlockStoreFs extends BlockStoreBase {
         return P.join(
```
What about this P.join?

guymguym reviewed (changes requested):
@liranmauda I'm not sure what this currently includes.
I see more P.try and P.join left behind.
Why is that?

guymguym commented:
@liranmauda What's the status? Is it ready?

liranmauda commented:
@guymguym 
It can be reviewed. 
so there was some more `P.join` that I replaced and found some more `P.delay` that I have missed.
where bluebird is needed We wrapped it with a bluebird call.

P.resolve is still bluebird, we will hijack it Next PR when we will handle the timeout and defer (what do you think?)

guymguym commented:
OK I think P.resolve should be completely replaced to either Promise.resolve or Bluebird.resolve.

liranmauda commented:
/retest all

liranmauda commented:
/retest all

liranmauda commented:
/retest all

liranmauda commented:
/retest all

liranmauda commented:
/retest all
